src/iop/demosaic.c

```diff
@@ -974,15 +977,1208 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+    float (*const gmin)[TS] = (float(*)[TS])(buffer + TS * TS * (ndir * 3) * sizeof(float));
+    float (*const gmax)[TS] = (float(*)[TS])(buffer + TS * TS * (ndir * 3 + 1) * sizeof(float));
+    // homo and homosum reuse memory which is used earlier in the
+    // loop; each points to ndir single-channel TSxTS tiles
+    uint8_t (*const homo)[TS][TS] = (uint8_t(*)[TS][TS])(buffer + TS * TS * (ndir * 3) * sizeof(float));
+    uint8_t (*const homosum)[TS][TS]
+        = (uint8_t(*)[TS][TS])(buffer + TS * TS * (ndir * 3) * sizeof(float) + TS * TS * ndir * sizeof(uint8_t));
+    // append all fdc related buffers
+    float complex *fdc_buf_start = (float complex *)(buffer + TS * TS * (ndir * 4 + 3) * sizeof(float));
+    const int fdc_buf_size = TS * TS;
+#ifdef HAVE_FFTW3
+    float complex(*const i_src) = fdc_buf_start;
+#endif
+#ifndef HAVE_FFTW3
+    float(*const i_src) = (float *)fdc_buf_start;
+#endif
+    float complex(*const o_src) = fdc_buf_start + fdc_buf_size;
+#ifdef HAVE_FFTW3
+    float complex(*const i_kernel) = fdc_buf_start + fdc_buf_size * 2;
+    float complex(*const o_kernel) = fdc_buf_start + fdc_buf_size * 3;
+    float complex(*const C2mbuff) = fdc_buf_start + fdc_buf_size * 4;
+    float complex(*const C5mbuff) = fdc_buf_start + fdc_buf_size * 5;
+    float complex(*const C7mbuff) = fdc_buf_start + fdc_buf_size * 6;
+    // C10m are the last modulated coefficinets. By the time they are calculated, i_kernel becomes irrelevant.
+    float complex(*const C10mbuff) = i_kernel;
+#endif
+    // by the time the chroma values are calculated, o_src can be overwritten.
+    float(*const fdc_chroma) = (float *)o_src;
+
+    for(int left = -pad_tile; left < width - pad_tile; left += TS - (pad_tile * 2))
+    {
+      int mrow = MIN(top + TS, height + pad_tile);
+      int mcol = MIN(left + TS, width + pad_tile);
+#define CLIP(x) ((x < 0) ? 0.0 : (x > 1.0) ? 1.0 : x)
+
+      // Copy current tile from in to image buffer. If border goes
+      // beyond edges of image, fill with mirrored/interpolated edges.
+      // The extra border avoids discontinuities at image edges.
+      for(int row = top; row < mrow; row++)
+        for(int col = left; col < mcol; col++)
+        {
+          float(*const pix) = rgb[0][row - top][col - left];
+          if((col >= 0) && (row >= 0) && (col < width) && (row < height))
+          {
+            const int f = FCxtrans(row, col, roi_in, xtrans);
+            for(int c = 0; c < 3; c++) pix[c] = (c == f) ? in[roi_in->width * row + col] : 0.f;
+#ifdef HAVE_FFTW3
+            *(i_src + TS * (row - top) + (col - left)) = CLIP(in[roi_in->width * row + col])
+                                                         + 0.f * _Complex_I; // here is the problem
+#endif
+#ifndef HAVE_FFTW3
+            *(i_src + TS * (row - top) + (col - left)) = CLIP(in[roi_in->width * row + col]); // here is the
+                                                                                              // problem
```

Comment:
@LebedevRI I guess the comments here refer to the green edge issue. Any case, these comments are not helpful without context and your workaround works well. Does the issue need to be investigated upstream though (rawspeed)?

Reply:
Please provide the sample with the problem.

Reply:
(it's impossible to say anything without seeing exactly that same sample where you have observed the issue.)

Reply:
It appears that I get values out of the expected range [0.f ... 1.f] back from 'in'. If you take the raw file [here](https://www.dropbox.com/s/o26gp98nwuli9qn/green_border_problem.RAF?dl=0) and process it with fdc demosaicking and remove the CLIPs, then you will see a green border on the right hand side. This is only for X-T2 images, X-Pro 1 images don't have this issue. 
The following code also remedies the issue:
`#ifdef HAVE_FFTW3
            *(i_src + TS * (row - top) + (col - left)) = ( (in[roi_in->width * row + col] < 0.f) ? 0.f : in[roi_in->width * row + col] )
                                                         + 0.f * _Complex_I;
#endif`
Hence my conclusion that _negative_ values must be the problem. I did the same check for greater than one, for isna() and for isinf(), without results. So negative values appear to be the reason.

Reply:
I'm gonna bet the crop is wrong. Will check later.

Reply:
As to the question of upstream, I am not sufficiently the expert to know what processing sits between rawspeed itself and the data at this point, but I'll try to investigate further.

Reply:
Thanks @LebedevRI .

Reply:
Yep, pretty clearly, i'm right, will adjust.

Reply:
Cool, thanks a lot. Then I'll remove the CLIPs from my side.

Reply:
Pushed, hopefully that fixed the issue you were observing. Please remove that clipping.

Reply:
OK, I synced my fork back to upstream, taking your commit dc32f19 into account. I removed the CLIPping. Bad news: the issue isn't fully gone. Green border on the right hand side is still there, unfortunately.

Reply:
Hold on, my apologies. My bad. My eyes had been looking at too much code yesterday ;-)
`<Crop x="0" y="0" width="-132" height="0"/>`
did the trick! 1000times thanks.